aten/src/ATen/native/UpSampleNearest2d.cpp

```diff
@@ -3,44 +3,79 @@
 #include <ATen/native/UpSample.h>
 
 namespace at {
-namespace native {
-namespace {
+namespace meta {
 
-static void upsample_nearest2d_out_cpu_template(
-    Tensor& output,
-    const Tensor& input,
-    IntArrayRef output_size,
-    c10::optional<double> scales_h,
-    c10::optional<double> scales_w) {
+static std::array<int64_t, 4> upsample_nearest2d_common_check(IntArrayRef input_size, IntArrayRef output_size) {
   TORCH_CHECK(
       output_size.size() == 2,
-      "It is expected output_size equals to 2, but got size ",
+      "It is expected output_size equals to 1, but got size ",
       output_size.size());
 
-  int64_t output_height = output_size[0];
-  int64_t output_width = output_size[1];
+  TORCH_CHECK(
+      input_size.size() == 4,
+      "It is expected input_size equals to 3, but got size ",
+      input_size.size());
 
-  int64_t nbatch = input.size(0);
-  int64_t channels = input.size(1);
-  int64_t input_height = input.size(2);
-  int64_t input_width = input.size(3);
+  int64_t output_width = output_size[0];
+  int64_t output_height = output_size[1];
 
-  upsample_2d_shape_check(
-      input,
-      Tensor(),
-      nbatch,
-      channels,
+  int64_t nbatch = input_size[0];
+  int64_t channels = input_size[1];
+  int64_t input_width = input_size[2];
+  int64_t input_height = input_size[3];
+
+  TORCH_CHECK(
+      input_height > 0 && input_width > 0 && output_height > 0 &&
+          output_width > 0,
+      "Input and output sizes should be greater than 0,"
+      " but got input (H: ",
       input_height,
+      ", W: ",
       input_width,
+      ") output (H: ",
       output_height,
-      output_width);
+      ", W: ",
+      output_width,
+      ")");
+
+  return {nbatch, channels, output_width, output_height};
+}
 
-  output.resize_({nbatch, channels, output_height, output_width}, input.suggest_memory_format());
+TORCH_META_FUNC(upsample_nearest2d) (
+  const Tensor& input, IntArrayRef output_size, c10::optional<double> scales_h, c10::optional<double> scales_w
+) {
+  auto full_output_size = upsample_nearest2d_common_check(input.sizes(), output_size);
 
-  AT_ASSERT(input_width > 0 && output_width > 0);
-  upsample_nearest2d_kernel(kCPU, output, input, scales_h, scales_w);
+  // Allow for empty batch size but not other dimensions
+  TORCH_CHECK(
+      (input.numel() != 0 ||
+        (input.size(1) != 0 && input.size(2) != 0 && input.size(3) != 0)
+        ) &&
+      input.dim() == 4,
```

Comment:
you've just checked that input.dim() == 4, this is always true, and if it weren't, there would be trouble in input.size(3)

Reply:
hmm, this might actually be a useful check because input.size(3) might've been short-circuited if input.numel() != 0. Either way if we use prod_intlist method, we will have to keep this check.

Reply:
`upsample_nearest_2d_common_check` checks that `input_size.size() == 4`, so you can get to this line only if input.dim == 4, no need to check for it again.